Streaming inputs for 5.0

freddyaboulton opened the pull request:
## Description
Reworks how input streams are handled by the queue. All the data for a stream is processed as the same queue event as opposed to separate events for each chunk of data. This is a smoother experience as the queue won't switch between streams until the stream is closed.

Since streams can theoretically run forever, streaming input events now have a `time_limit` parameter to control for how long a given stream will be processed by the queue. So `time_limit=10` means that a user will "own" the queue worker processing that stream for 10 seconds. All others will have to wait. To handle streams concurrently, the `concurrency_count` in that event needs to be set to be greater than 1.

Also added a `stream_frequency` parameter to `Image` and `Audio` components that controls the stream sampling rate. Its unit is Hz (so `stream_frequency=10` means 10 chunks per second)/

### Demo (stream image in)

```python
import gradio as gr

import gradio as gr
import numpy as np
import cv2


def transform_cv2(frame, transform):
    if transform == "cartoon":
        # prepare color
        img_color = cv2.pyrDown(cv2.pyrDown(frame))
        for _ in range(6):
            img_color = cv2.bilateralFilter(img_color, 9, 9, 7)
        img_color = cv2.pyrUp(cv2.pyrUp(img_color))

        # prepare edges
        img_edges = cv2.cvtColor(frame, cv2.COLOR_RGB2GRAY)
        img_edges = cv2.adaptiveThreshold(
            cv2.medianBlur(img_edges, 7),
            255,
            cv2.ADAPTIVE_THRESH_MEAN_C,
            cv2.THRESH_BINARY,
            9,
            2,
        )
        img_edges = cv2.cvtColor(img_edges, cv2.COLOR_GRAY2RGB)
        # combine color and edges
        img = cv2.bitwise_and(img_color, img_edges)
        return img
    elif transform == "edges":
        # perform edge detection
        img = cv2.cvtColor(cv2.Canny(frame, 100, 200), cv2.COLOR_GRAY2BGR)
        return img
    else:
        return np.flipud(frame)

with gr.Blocks() as demo:
    gr.Markdown("# New Image Streaming In/Out 📸")
    with gr.Row():
        with gr.Column():
            transform = gr.Dropdown(choices=["cartoon", "edges", "flip"],
                                    value="flip", label="Transformation")
            input_img = gr.Image(sources=["webcam"], stream_frequency=10, type="numpy")
        with gr.Column():
            output_img = gr.Image(format="jpeg")
        dep = input_img.stream(transform_cv2, [input_img, transform], [output_img], time_limit=25)

if __name__ == "__main__":
    demo.launch()
```
![stream_image_in_out](https://github.com/user-attachments/assets/e06bb4de-3862-42df-bc98-6603fcd3053b)

### Demo (audio in)

```python
import gradio as gr

from dotenv import load_dotenv
load_dotenv()

from openai import OpenAI

client = OpenAI()

def transcribe(audio: str, state: str):
    next_chunk = client.audio.transcriptions.create(
        model="whisper-1",
        file=open(audio, "rb")
    ).text
    state += " " + next_chunk
    return state, state

with gr.Blocks() as demo:
    gr.Markdown("# New Audio Streaming In 🎤")
    inp = gr.Audio(sources=["microphone"], stream_frequency=1, type="filepath")
    out = gr.Text()
    transcription = gr.State(value="")
    clear = gr.Button("Clear")

    inp.stream(transcribe, [inp, transcription], [out, transcription], time_limit=10)

if __name__ == "__main__":
    demo.launch()
```


## 🎯 PRs Should Target Issues

Closes: #5873 
Closes: #5848

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

freddyaboulton commented on js/app/src/Blocks.svelte:
```diff
@@ -254,25 +255,38 @@
 		}
 
-		async function make_prediction(payload: Payload): Promise<void> {
+		async function make_prediction(
```
If we're streaming, `make_prediction` will send the latest stream event data to the queue as opposed to creating a whole new event.

freddyaboulton commented on js/app/src/Blocks.svelte:
```diff
@@ -518,6 +535,17 @@
 				messages = [new_message(data, -1, event), ...messages];
 			} else if (event == "clear_status") {
 				update_status(id, "complete", data);
+			} else if (event == "close_stream") {
```
You'll see `close_stream` in two places in the front-end - 

1. The first is here where I've added a `close_stream` event that can be emitted by the svelte components. This signals to the queue that the stream has been closed client-side and so we can just terminate the event running on the queue
2. The second is the `close_stream` function in `init.ts`. If the stream is closed server-side (time limit reached), we need to signal to the components that they should stop streaming. Each streaming component exposes a `close_stream` callback that the app will call once the stream is closed server-side

freddyaboulton commented on js/image/shared/ImageUploader.svelte:
```diff
@@ -47,7 +50,7 @@
 		pending = true;
 		const f = await upload_input.load_files([
-			new File([img_blob], `webcam.png`)
+			new File([img_blob], `image/${streaming ? "jpeg" : "png"}`)
```
We should use `jpeg` for streaming since they're smaller

freddyaboulton commented on gradio/events.py:
```diff
@@ -348,6 +348,7 @@ def __init__(
         trigger_after: int | None = None,
         trigger_only_on_success: bool = False,
         doc: str = "",
+        connection: Literal["sse", "stream"] = "sse",
```
`connection` is maybe a confusing name here, but basically we need to be able to tell which events are streaming events without relying on the event name

abidlabs replied:
Why not just a boolean `streaming` parameter?

freddyaboulton commented on gradio/exceptions.py:
```diff
@@ -59,6 +59,12 @@ class GradioVersionIncompatibleError(Exception):
     pass
 
 
+class StreamingTimeoutError(Exception):
```
I think I can remove this

freddyaboulton commented on gradio/processing_utils.py:
```diff
@@ -489,6 +489,8 @@ async def _move_to_cache(d: dict):
 
 def add_root_url(data: dict | list, root_url: str, previous_root_url: str | None):
     def _add_root_url(file_dict: dict):
+        if file_dict.get("data") is not None:
```
Will remove - typo

freddyaboulton commented on gradio/queueing.py:
```diff
@@ -532,15 +597,25 @@ async def process_events(
                 err = None
+                for event in awake_events:
+                    event.run_time += end - start
+                    event.n_calls += 1
```
Will remove this n_calls line

abidlabs commented on gradio/components/audio.py:
```diff
@@ -109,6 +109,7 @@ def __init__(
         max_length: int | None = None,
         waveform_options: WaveformOptions | dict | None = None,
         loop: bool = False,
+        stream_frequency: float = 2,
```
Nitty nit, but we could call this `stream_every` to make it consistent with our other `every` parameter

abidlabs replied:
Could we consolidate this and the `streaming` arguments?

abidlabs replied:
Actually could we make this a parameter of .stream()? That way (a) different events could have different frequencies (b) this parameter and the time_limit() parameter are in the same place (c) if someone isn’t using streaming, they don’t have to think about this parameter

freddyaboulton replied:
Done

abidlabs replied:
> Actually could we make this a parameter of .stream()? That way (a) different events could have different frequencies (b) this parameter and the time_limit() parameter are in the same place (c) if someone isn’t using streaming, they don’t have to think about this parameter

Thoughts on this @freddyaboulton?

freddyaboulton replied:
Yes it's something I thought about but it's a bit awkward since we don't have a way of passing dependency props into components but regardless I will implement.

abidlabs commented on gradio/components/image.py:
```diff
@@ -101,6 +102,7 @@ def __init__(
             key: if assigned, will be used to assume identity across a re-render. Components that have the same key across a re-render will have their value preserved.
             mirror_webcam: If True webcam will be mirrored. Default is True.
             show_share_button: If True, will show a share icon in the corner of the component that allows user to share outputs to Hugging Face Spaces Discussions. If False, icon does not appear. If set to None (default behavior), then the icon appears if this Gradio app is launched on Spaces, but not otherwise.
+            stream_frequency: The number of times per second to stream the webcam feed. Only valid if streaming is True.
```
Is it true that this argument is valid only if streaming is True? In your example demo, you have this line:

>             input_img = gr.Image(sources=["webcam"], stream_frequency=10, type="numpy")

abidlabs replied:
I'm a little confused about the `streaming` argument actually. Is it used at all for input streaming? If I understand correctly, it is only used to set the behavior when a component is being used as output. We should make this clearer in the docs and docstrings.

freddyaboulton replied:
Yes `streaming` parameter is only used for outputs. Will make a note of that in docstring.

abidlabs replied:
We do have this logic:

```
            self.sources = (
                ["webcam"] if streaming else ["upload", "webcam", "clipboard"]
            )
```

but I think we can change this to: if an image component only has .stream() events, then the default source is webcam

abidlabs replied:
Perhaps we should rename the `streaming` parameter to `stream_out` to help users know which flags they need to set for which use case (if we are worried about breaking, then we could at least deprecate the former and assign its value to the latter)

abidlabs commented:
Very nice @freddyaboulton! I took it for a spin and the streaming works very nicely. I left a couple of minor points above, but my main feedback is around the `time_limit` experience. 

(1) I mentioned this before, but when you're using the app locally (or in a private Space), the "Time limit reached!" modal is jarring as it seems like the Gradio app is putting an artificial limit in my ability to interact with my own model / demo. I think the simplest way would be to automatically resend the stream event when it is complete to the end of the queue, which would work no matter how many people are using the app. However, this might produce an interruption between consecutive events, so there might be a better way to manage the single-user use case

(2) I like the orange bar that partially disappears to indicate the streaming time. However, I'm not sure if it makes sense to put it in the output component, since the `time_limit` is actually more connected to the input component. In some cases, you might have multiple output components (or _no_ output components): 

<img width="871" alt="image" src="https://github.com/user-attachments/assets/3dc4f5fb-9d37-4cae-96b4-921b2bac4a78">

I'd suggest moving this indicator over to the input component. Perhaps you could have a red border around the record button which partially disappears to indicate the time limit.

abidlabs commented:
Thanks @freddyaboulton changes look good! The single-user experience is quite good, but a couple of suggestions to improve the streaming input UX when there are multiple users using the demo at the same time:

1. When you are waiting for your turn in the queue, there is no progress indicator in the top right of the output to let you know your position in the queue, eta, etc. like there normally is for events:

<img width="1250" alt="image" src="https://github.com/user-attachments/assets/633f5afe-783c-4258-b18f-626894c96ed2">

2. Part of the progress indicator is hidden behind the label, which can be misleading. I'd suggest either putting it on the bottom of the component (closer to the record button), or making the progress indicator a circular border around the record button which gradually disappears (I think this UX would look really nice). Nit: regardless of which way we go with, can we color it red? Visually it would be more connected to the record button that way

<img width="629" alt="image" src="https://github.com/user-attachments/assets/5101ce44-925d-4022-90ec-447891b65306">



This is the demo that I used to test (opened the demo in multiple tabs):

```py
import gradio as gr

import gradio as gr
import numpy as np
import cv2


def transform_cv2(frame, transform):
    if transform == "cartoon":
        # prepare color
        img_color = cv2.pyrDown(cv2.pyrDown(frame))
        for _ in range(6):
            img_color = cv2.bilateralFilter(img_color, 9, 9, 7)
        img_color = cv2.pyrUp(cv2.pyrUp(img_color))

        # prepare edges
        img_edges = cv2.cvtColor(frame, cv2.COLOR_RGB2GRAY)
        img_edges = cv2.adaptiveThreshold(
            cv2.medianBlur(img_edges, 7),
            255,
            cv2.ADAPTIVE_THRESH_MEAN_C,
            cv2.THRESH_BINARY,
            9,
            2,
        )
        img_edges = cv2.cvtColor(img_edges, cv2.COLOR_GRAY2RGB)
        # combine color and edges
        img = cv2.bitwise_and(img_color, img_edges)
        return img
    elif transform == "edges":
        # perform edge detection
        img = cv2.cvtColor(cv2.Canny(frame, 100, 200), cv2.COLOR_GRAY2BGR)
        return img
    else:
        return np.flipud(frame)

with gr.Blocks() as demo:
    gr.Markdown("# New Image Streaming In/Out 📸")
    with gr.Row():
        with gr.Column():
            transform = gr.Dropdown(choices=["cartoon", "edges", "flip"],
                                    value="flip", label="Transformation")
            input_img = gr.Image(sources=["webcam"], stream_every=0.1, type="numpy")
        with gr.Column():
            output_img = gr.Image(format="jpeg")
        dep = input_img.stream(transform_cv2, [input_img, transform], [output_img], time_limit=25)

if __name__ == "__main__":
    demo.launch()
```

abidlabs commented:
Btw similar to (1), when your turn is up in the queue, it should again show the progress indicator with the queue position and eta to let you know when it would be your turn again, as otherwise you're in this intermediate state where it seems like the demo has just stopped working:

<img width="1248" alt="image" src="https://github.com/user-attachments/assets/1f6f63d9-d9d9-4acc-93e4-0d4f50fc126c">

aliabid94 commented:
I tried running the audio streaming demo drom PR desc and got:
```
Traceback (most recent call last):
  File "/Users/aliabid/Desktop/gradio/demo/hello_world/run.py", line 1, in <module>
    import gradio as gr
  File "/Users/aliabid/Desktop/gradio/gradio/__init__.py", line 3, in <module>
    import gradio._simple_templates
  File "/Users/aliabid/Desktop/gradio/gradio/_simple_templates/__init__.py", line 1, in <module>
    from .simpledropdown import SimpleDropdown
  File "/Users/aliabid/Desktop/gradio/gradio/_simple_templates/simpledropdown.py", line 6, in <module>
    from gradio.components.base import Component, FormComponent
  File "/Users/aliabid/Desktop/gradio/gradio/components/__init__.py", line 1, in <module>
    from gradio.components.annotated_image import AnnotatedImage
  File "/Users/aliabid/Desktop/gradio/gradio/components/annotated_image.py", line 14, in <module>
    from gradio.components.base import Component
  File "/Users/aliabid/Desktop/gradio/gradio/components/base.py", line 20, in <module>
    from gradio.blocks import Block, BlockContext
  File "/Users/aliabid/Desktop/gradio/gradio/blocks.py", line 39, in <module>
    from gradio import (
  File "/Users/aliabid/Desktop/gradio/gradio/networking.py", line 14, in <module>
    from gradio.routes import App  # HACK: to avoid circular import # noqa: F401
  File "/Users/aliabid/Desktop/gradio/gradio/routes.py", line 89, in <module>
    from gradio.server_messages import (
  File "/Users/aliabid/Desktop/gradio/gradio/server_messages.py", line 51, in <module>
    class ProcessGeneratingMessage(BaseMessage):
  File "/Users/aliabid/Desktop/gradio/gradio/server_messages.py", line 52, in ProcessGeneratingMessage
    msg: Literal[ServerMessage.process_generating, ServerMessage.process_streaming] = (  # type: ignore
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/enum.py", line 429, in __getattr__
    raise AttributeError(name) from None
AttributeError: process_streaming
```

abidlabs commented:
@aliabid94 make sure you're installing the client from this PR as well. I had the same issue

`pip install -e client/python`

freddyaboulton commented:
@abidlabs I believe I've addressed all of your comments:

* stream_every moved to event definition as opposed to component constructor
* time limit indicator moved to the bottom of the component
* queue status displayed now for streaming events

Additionally, the record button icon for images now goes off of the primary theme color to match the audio component!

abidlabs commented:
@freddyaboulton I'm seeing a weird issue which is that when I run this demo:

```py
import gradio as gr

import gradio as gr
import numpy as np
import cv2


def transform_cv2(frame, transform):
    if transform == "cartoon":
        # prepare color
        img_color = cv2.pyrDown(cv2.pyrDown(frame))
        for _ in range(6):
            img_color = cv2.bilateralFilter(img_color, 9, 9, 7)
        img_color = cv2.pyrUp(cv2.pyrUp(img_color))

        # prepare edges
        img_edges = cv2.cvtColor(frame, cv2.COLOR_RGB2GRAY)
        img_edges = cv2.adaptiveThreshold(
            cv2.medianBlur(img_edges, 7),
            255,
            cv2.ADAPTIVE_THRESH_MEAN_C,
            cv2.THRESH_BINARY,
            9,
            2,
        )
        img_edges = cv2.cvtColor(img_edges, cv2.COLOR_GRAY2RGB)
        # combine color and edges
        img = cv2.bitwise_and(img_color, img_edges)
        return img
    elif transform == "edges":
        # perform edge detection
        img = cv2.cvtColor(cv2.Canny(frame, 100, 200), cv2.COLOR_GRAY2BGR)
        return img
    else:
        return np.flipud(frame)

with gr.Blocks() as demo:
    gr.Markdown("# New Image Streaming In/Out 📸")
    with gr.Row():
        with gr.Column():
            transform = gr.Dropdown(choices=["cartoon", "edges", "flip"],
                                    value="flip", label="Transformation")
            input_img = gr.Image(sources=["webcam"], stream_every=0.1, type="numpy")
        with gr.Column():
            output_img = gr.Image(format="jpeg")
        dep = input_img.stream(transform_cv2, [input_img, transform], [output_img], time_limit=25)

if __name__ == "__main__":
    demo.launch()
```

on three separate tabs, the third demo never runs. It just freezes up in this state:

<img width="1347" alt="image" src="https://github.com/user-attachments/assets/16e0936f-db02-481a-a32d-d7810fbaa4b2">

Sometimes the demo will run, but it will be very laggy. If you switch back to the first and second demos, they also start to freeze up or not work. The error-prone behavior varies a little bit, but I consistently get some sort of error when I open up the demo on three tabs (but not on just two). 

I also see a bunch of errors like this in the console:


<img width="384" alt="image" src="https://github.com/user-attachments/assets/3ff661f9-41bd-4aae-a0ba-61b15a3b7e0d">

Besides that, two small suggestions:
* Can we set `show_progress="minimal"` as the default for `.stream()` events? I believe we do the same thing for `.change()`
* The queue calculation seems slightly off. When I run two events with `stream_every=10`, the queue calculation should show an ETA of 10 seconds, but it actually shows 20, because its adding the total streaming time of the current job, which it shouldn't since users would want to know the time until the *start* of the stream

abidlabs commented:
Discussed with @freddyaboulton and the reason I'm seeing the error is because the browser is throttling the messages. Won't happen with different browsers so I think its ok

abidlabs commented on gradio/component_meta.py:
```diff
@@ -57,6 +59,8 @@ def {{ event }}(self,
             concurrency_id: if set, this is the id of the concurrency group. Events with the same concurrency_id will be limited by the lowest set concurrency_limit.
             show_api: whether to show this event in the "view API" page of the Gradio app, or in the ".view_api()" method of the Gradio clients. Unlike setting api_name to False, setting show_api to False will still allow downstream apps as well as the Clients to use this event. If fn is None, show_api will automatically be set to False.
+            time_limit: The time limit for the function to run. Only applies if connection is "stream".
+            stream_every: The latency (in seconds) at which stream chunks are sent to the backend. Defaults to 0.5 seconds. Only applies for "stream" events.
```
```suggestion
            stream_every: The latency (in seconds) at which stream chunks are sent to the backend. Defaults to 0.5 seconds. Parameter only used for the `.stream()` event.
```

abidlabs replied:
wording nit, applies to the other docstrings as well

abidlabs commented on gradio/component_meta.py:
```diff
@@ -57,6 +59,8 @@ def {{ event }}(self,
             concurrency_limit: if set, this is the maximum number of this event that can be running simultaneously. Can be set to None to mean no concurrency_limit (any number of this event can be running simultaneously). Set to "default" to use the default concurrency limit (defined by the `default_concurrency_limit` parameter in `Blocks.queue()`, which itself is 1 by default).
             concurrency_id: if set, this is the id of the concurrency group. Events with the same concurrency_id will be limited by the lowest set concurrency_limit.
             show_api: whether to show this event in the "view API" page of the Gradio app, or in the ".view_api()" method of the Gradio clients. Unlike setting api_name to False, setting show_api to False will still allow downstream apps as well as the Clients to use this event. If fn is None, show_api will automatically be set to False.
+            time_limit: The time limit for the function to run. Only applies if connection is "stream".
```
```suggestion
            time_limit: The time limit for the function to run. Parameter only used for the `.stream()` event.
```

abidlabs commented:
Thanks @freddyaboulton, gave it another review. I think everything looks good to me, except for three small things that I noticed. Two I've mentioned above, will copy here:

* Can we set show_progress="minimal" as the default for .stream() events? I believe we do the same thing for .change()


* The queue calculation seems slightly off. When I run two events with stream_every=10, the queue calculation should show an ETA of 10 seconds, but it actually shows 20, because its adding the total streaming time of the current job, which it shouldn't since users would want to know the time until the start of the stream


The last thing I noticed is that when using `gr.Audio` as an input with streaming, it seems that even though the stream event gets restarted after the streaming is done, you don't see the red bar reappear. Here's the demo I am running:

```py
import gradio as gr

t = "the big blue " * 5
i = 0

def transcribe(audio: str, state: str):
    global i
    next_chunk = t[i:i+1]
    i += 1
    state += " " + next_chunk
    return state, state

with gr.Blocks() as demo:
    gr.Markdown("# New Audio Streaming In 🎤")
    inp = gr.Audio(sources=["microphone"], type="filepath")
    out = gr.Text()
    transcription = gr.State(value="")
    clear = gr.Button("Clear")

    inp.stream(transcribe, [inp, transcription], [out, transcription], time_limit=10, stream_every=1)

if __name__ == "__main__":
    demo.launch()
```

and you can see after 10 seconds, it restarts and continues streaming, but there's no red bar:

<img width="808" alt="image" src="https://github.com/user-attachments/assets/3ad7ad29-ca7f-41b8-93d1-90faf229832b">

freddyaboulton commented:
Thanks @abidlabs ! I made those changes. We don't set `show_progress='minimal'` for change events but I did it for streaming here anyways. I modified the queue eta for streaming and fixed progress tracker issue with the audio component.

abidlabs reviewed (approved):
Awesome PR @freddyaboulton, and thanks for addressing all the suggestions. Everything looks good to me. 

It'd be great to add some documentation with real-world examples, but we can do that in a subsequent PR

freddyaboulton commented:
Thank you so much for the reviews!
